### PR TITLE
st1 [2054][IMP] report_common_nrq, report_purchase_nrq, report_sale_nrq: unit price formatting

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[2054
](https://www.quartile.co/web#id=2054&action=771&model=project.task&view_type=form&menu_id=505)

Parent task:[1896](https://www.quartile.co/web#view_type=form&model=project.task&id=1896&active_id=1896&menu_id=517)

### CODE REVIEW

File: report_purchase_nrq/report/purchase_report_template.xml [hunk 1]
```diff
@@ -53,7 +53,16 @@
                                 />
                             </td>
                             <td class="text-right">
-                                <span t-field="line.price_unit" />
+                                <t t-set="price_unit" t-value="line.price_unit" />
+                                <t t-if="round(price_unit) == price_unit">
+                                    <span
+                                        t-esc="float('%.2f'%(price_unit))"
+                                        t-options='{"widget": "monetary", "display_currency": doc.currency_id, "precision":2}'
```
Comment (yostashiro) [inline report_purchase_nrq/report/purchase_report_template.xml]: Why should this this `t-options` be here?

Comment (TomohiroAbe-Quartile) [reply report_purchase_nrq/report/purchase_report_template.xml]: Now t-esc is float, so if t-option not be here, currency symbol and comma(,) will not appear.

Comment (yostashiro) [reply report_purchase_nrq/report/purchase_report_template.xml]: > Now t-esc is float, so if t-option not be here, currency symbol and comma(,) will not appear.

It makes the field presentation inconsistent whether there is something below decimal or not. `price_unit` is a float field in the first place (i.e. currently no currency symbol shows).

Get rid of `t-options`, and try replacing
```
t-esc="float('%.2f'%(price_unit))"
```
with
```
t-esc="'%:,.0f'%(price_unit)"
```
to see if it works?

Please also double-check the assumption - my understanding is that we are supposed to change the decimal precision setting of 'Product Price' from `0` to `2`.

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: report_purchase_nrq/report/purchase_report_template.xml (hunk 1 above)
Comment (TomohiroAbe-Quartile) [reply report_purchase_nrq/report/purchase_report_template.xml]: @yostashiro 
It not works.
If you can, please check how it works with comma.

I have not found how to do yet.

Comment (yostashiro) [reply report_purchase_nrq/report/purchase_report_template.xml]: @TomohiroAbe-Quartile Take a look here: https://github.com/qrtl/rpl-private/pull/209/commits/3d75df04a555f4de1f58888483503901fea34270#diff-0e1826db127901de699b3b33bddebec4ece23baf5430c9e3a241dd65c10781f1R36-R37

Comment (TomohiroAbe-Quartile) [reply report_purchase_nrq/report/purchase_report_template.xml]: Thank you, it works as expected. I will learn why this line does working.

**Comment by TomohiroAbe-Quartile** [comment]
I deleted t-if conditional branch below:
`round(price_unit) == price_unit `

I have no idea to use this line.

**Comment by TomohiroAbe-Quartile** [comment]
@yostashiro 
I edited t-if conditional branch that added round(price_unit).
And add in report_common_nrq and report_sale_nrq.
report_account_nrq doesn't have prie_unit, so I think this is out of scope in this PR.

File: report_common_nrq/report/common_template.xml [hunk 2]
```diff
@@ -474,7 +474,21 @@
...
+                                    >
+                                        <span t-esc="'{:,.2f}'.format(price_unit)" />
+                                    </t>
+                                    <t
+                                        t-if="round(price_unit) == price_unit
+                                        and
+                                        doc.currency_id.rounding >= 1"
+                                    >
+                                        <span t-esc="'{:,.0f}'.format(price_unit)" />
+                                    </t>
```
Comment (yostashiro) [inline report_common_nrq/report/common_template.xml]: @TomohiroAbe-Quartile Can we make a template for this and call it with t-call in respective templates?

Comment (TomohiroAbe-Quartile) [reply report_common_nrq/report/common_template.xml]: I made a "report_common_nrq.price_unit_display" template.

File: report_common_nrq/report/common_template.xml [hunk 3]
```diff
@@ -354,6 +354,23 @@
             </tr>
         </table>
     </template>
+    <template id="price_unit_display">
+        <t t-set="price_unit" t-value="l.price_unit" />
```
Comment (yostashiro) [inline report_common_nrq/report/common_template.xml]: We should probably leave this line outside of the template to leave the flexibility on what should be the value of `price_unit` (pay attention to the lines I highlighted in my last review).

File: report_common_nrq/report/common_template.xml [hunk 4]
```diff
@@ -474,7 +491,21 @@
...
+                                    >
+                                        <span t-esc="'{:,.2f}'.format(price_unit)" />
+                                    </t>
+                                    <t
+                                        t-if="round(price_unit) == price_unit
+                                        and
+                                        doc.currency_id.rounding >= 1"
+                                    >
+                                        <span t-esc="'{:,.0f}'.format(price_unit)" />
+                                    </t>
```
Comment (yostashiro) [inline report_common_nrq/report/common_template.xml]: Any reason why you are not calling the template here?

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: report_common_nrq/report/common_template.xml (hunk 3 above)
Comment (TomohiroAbe-Quartile) [reply report_common_nrq/report/common_template.xml]: Thank you. I see.
I added comment to define t-set attr.

File: report_common_nrq/report/common_template.xml (hunk 4 above)
Comment (TomohiroAbe-Quartile) [reply report_common_nrq/report/common_template.xml]: Sorry, nothing to reason.

File: report_common_nrq/report/common_template.xml [hunk 5]
```diff
@@ -354,6 +354,27 @@
             </tr>
         </table>
     </template>
+    <template id="price_unit_value_format">
+        <!--
```
Comment (tack-yabeQuartile) [inline report_common_nrq/report/common_template.xml:358]: What is definition of "price_unit"? (Where is line of definition?) 
I could not find it in this report_common_nrq module. 
If the definition is written in depends, you can write as "t-value="DependingModuleName.price_unit".
If not, you must define "price_unit" variable/field in report_common_nrq/models/*.py
・depends 
"account", "sale", "purchase", "base_partner_number"

File: report_common_nrq/report/common_template.xml [hunk 6]
```diff
@@ -354,6 +354,27 @@
...
+            It should define the price_unit value.
+            e.g.
+            <t t-set="price_unit" t-value="l.price_unit" />
```
Comment (tack-yabeQuartile) [inline report_common_nrq/report/common_template.xml:361]: What is "l" in this template? This is not defined in the same template.
Maybe you use "l in doc.invoice_line_ids" in L383(before change), but this is defined in another template.

File: report_common_nrq/report/common_template.xml (hunk 5 above)
Comment (TomohiroAbe-Quartile) [reply report_common_nrq/report/common_template.xml:358]: You can see this line.
https://github.com/qrtl/nrq-custom/blob/97b84fc32ba9a8807cae8f308b0312c938bd8cfc/report_purchase_nrq/report/purchase_report_template.xml#L56

price_unit is an item in which the price_unit column has been replaced by t-as.

The definition name of the model instance differs depending on the situation. line, l and etc.

[So this template isn't able to contain price_unit value. (ytashiro left comment.)](https://github.com/qrtl/nrq-custom/pull/223#discussion_r755878317)

Comment (tack-yabeQuartile) [reply report_common_nrq/report/common_template.xml:358]: I could not understand the situation enough.
It seems that you have not inherited the definition of price_unit or l in report_purchase_nrq to this report_common_nrq module. 
Why don't you do that?  Does the template (id="price_unit_value_format") work intendedly?

**Comment by TomohiroAbe-Quartile** [comment]
@tack-yabeQuartile 
Point: report_purchase_nrq and report_sale_nrq (each modules) depend report_common_nrq.
And each modules call template on themselves.

So  each modules are children of report_common_nrq, and defines a model instance such as line, l, etc by t-as,  for each modules.

**Review by tack-yabeQuartile — APPROVED** [review]
I understand the code.
LGTM.

**Review by yostashiro — APPROVED** [review]

**Comment by TomohiroAbe-Quartile** [comment]
I confirmed reports below:
Sale and Purchase: account_invoice
purchase_order
sale_quotation

**Pull request merged**